### ISSUE TITLE
Build errors when integrating Branch using SPM in Xcode 12.5b3 when Warnings are treated as Errors

### ISSUE THREAD

**Issue by gransell** [BranchMetrics/ios-branch-deep-linking-attribution#1101]
Hi

When building in Xcode 12.5b3 we get the following build errors because of missing .h files in `Branch.h`

```
Showing Recent Messages
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BNCTuneUtility.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BNCKeyChain.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BNCTelephony.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BranchCPIDRequest.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BNCAppleReceipt.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BranchContentPathProperties.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BranchCreditHistoryRequest.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BNCSKAdNetwork.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BNCJSONUtility.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BNCAppleAdClient.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BNCAppGroupsData.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BNCFieldDefines.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BNCSpotlightService.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BranchRedeemRewardsRequest.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BNCCallbackMap.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BranchShortUrlSyncRequest.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BranchOpenRequest.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BranchShortUrlRequest.h'
/Users/gransell/Library/Developer/Xcode/DerivedData/QapitalApp-dmxntnlexxpmpzglcjqigbzxmirt/SourcePackages/checkouts/ios-branch-deep-linking-attribution/Branch-SDK/Branch.h:1906:1: Umbrella header for module 'Branch' does not include header 'BranchLogoutRequest.h'
```

**Comment by echo-branch** [comment]
@gransell 
Were you able to get past this issue?  I wasn't able to reproduce the issue.

**Comment by gransell** [comment]
@echo-branch 
Thanks for looking into this.
I just gave this another look with a clean test project importing only Branch and was not able to reproduce there. Could be something with our project. I'll keep investigating. I'll close this for now and report any findings here when we manage to fix it.

**Comment by gransell** [comment]
@echo-branch

I revisited this now and turns out we got the errors because we had "Treat warnings as errors" turned on. We still get the warnings in a simple Xcode project that imports branch via SPM and just does a simple `import Branch` in a swift file.

Attaching a sample project with the issue.
[BranchTest.zip](https://github.com/BranchMetrics/ios-branch-deep-linking-attribution/files/6336126/BranchTest.zip)

**Comment by echo-branch** [comment]
Ah yea, we have some clean up warnings tickets floating around for a while. We just haven't had the dev resources free up to fix them though.

**Comment by adamcarter93** [comment]
I'm getting this error too. A temporary fix was to also turn off "Treat warnings as errors".

**Comment by gransell** [comment]
Yeah. That's the "fix" we have opted for as well.

**Comment by james-lukanta-ck** [comment]
Hey Branch, is there any update on this? As you know, XCode 12.5 is officially out, and this is a blocker to upgrading to XCode 12.5 for many of us.

**Comment by echo-branch** [comment]
It's in progress, however we've found some of the warnings are non-trivial to remove and may take some time to refactor away.

**Comment by nikolouzos** [comment]
Hi @echo-branch! Any updates on this?

**Comment by echo-branch** [comment]
@nikolouzos 
The warnings should be fixed and merged into master, and will be in the next general release 1.40.0.
We're still working on some testing of server signature changes on that version, so you should not try it yet.

**Comment by jsorge** [comment]
Any update on this? I'd love to get back to warnings-as-errors but can't because of this issue.

**Comment by echo-branch** [comment]
@jsorge 
It's on track for the 1.40.0 release, however it did not make the 1.39.4 patch release.

**Comment by jsorge** [comment]
@echo-branch When will 1.40.0 land?

**Comment by rr-matthew-benjamin** [comment]
I am still seeing this issue when using `1.40.2`

**Comment by mdelamatata** [comment]
Same for `1.41.0`.

**Comment by mdelamatata** [comment]
Hi, this issue is present on `1.40.1` too.

**Comment by jf-branch** [comment]
Hi everyone, we are working on this and hope to get this out by the end of this month. We will update you all when the release is available.

**Comment by TheCoordinator** [comment]
It's quite annoying this issue, would be great if we can prioritise this and have it fixed in the new releases as it's been floating around for well over a year now

**Comment by echo-branch** [comment]
@TheCoordinator 
We hear you. Unfortunately the fix is to rename our library, which has cascading impact on all our build, test and documentation. It will get done, but it wasn't a simple one sprint fix we originally thought and will need to be on a major version update.

**Comment by jhurray** [comment]
@echo-branch totally understandable. It would be good to understand, from your perspective, why certain fixes are not viable 
- Adding the missing headers to Branch.h (understand we don't want files that aren't used but this could be a short term solution)
- Updating the modulemap file to exclude these headers explicitly (I don't see a modulemap for `Branch-SDK`) ([example](https://forums.swift.org/t/root-cause-of-umbrella-header-for-module-does-not-include/24063/2))

**Comment by echo-branch** [comment]
@jhurray 
So the error message is a bit confusing. The actual issue is that the `Branch.h` header is both the umbrella header and the header for a class. This leads to Xcode getting confused and throwing a lot of warnings. The fix is to create a dedicated umbrella header `BranchSDK.h` and use `Branch.h` as a class header. Internally we've already done this, but the impact outside the SDK is extensive.

For historical note, we ended up here by doing a quick fix hack to convert our static library SDK to a framework. That hack was to treat the `Branch.h` as both a class header and umbrella header.

**Comment by jhurray** [comment]
@echo-branch FYI I found a decent workaround if you want to share this w/ folks:
- Create a local swift package that wraps `Branch` (IMO always a good practice for 3rd party SDKs)
- in the `target` definition add a compiler flag to suppress objective-c warnings

```swift
            cSettings: [
                .unsafeFlags(["-suppress-warnings"])
            ]
``` 


Example package file

```swift
// swift-tools-version: 5.6
// The swift-tools-version declares the minimum version of Swift required to build this package.

import PackageDescription

let package = Package(
    name: "Foo",
    products: [
        .library(name: "Foo", targets: ["Foo"]),
    ],
    dependencies: [
        .package(url: "https://github.com/BranchMetrics/ios-branch-sdk-spm", from: "1.0.0"),
    ],
    targets: [
        .target(
            name: "Foo",
            dependencies: [
                .product(name: "Branch", package: "ios-branch-sdk-spm"),
            ],
            // Suppress warnings from `Branch` (50+ missing header warnings) related to their SDK structure
            // https://github.com/BranchMetrics/ios-branch-deep-linking-attribution/issues/1101
            // This can be removed when branch releases a new major version
            cSettings: [
                .unsafeFlags(["-suppress-warnings"])
            ]
        ),
        .testTarget(
            name: "FooTests",
            dependencies: [
                "Foo",
            ]
        ),
    ]
)
```

**Comment by pm-dev** [comment]
Any progress here?

**Comment by echo-branch** [comment]
@pm-dev 
We have a branch that's currently undergoing testing. Since it's somewhat disruptive we've had to hold it for other feature work a couple times already. Currently we're planning on shipping it after we ship an update for SKAN 4.0.

**Comment by AngeloGLobo** [comment]
These warnings are still present in 1.45.2. Any updates on when this will be released?

**Comment by victormarcias** [comment]
> @echo-branch How's the testing going? Any updates on when this is planned to be released? It's fairly annoying having so many warnings on each build!

+1

**Comment by victormarcias** [comment]
> 

Xcode 14. 1.45.2 - Warnings still there.

**Comment by echo-branch** [comment]
This should be corrected in 2.0.0. Let me know if it's still throwing warnings. 
Note, 2.0.0 changes the name of the framework from Branch to BranchSDK.
